go/interpreter/lfvm/instructions_test.go

```diff
@@ -1997,6 +1939,69 @@ func TestInstructions_Sha3_WritesCorrectHashInStack(t *testing.T) {
 	}
 }
 
+func TestInstructions_opLog(t *testing.T) {
+
+	tests := map[string]struct {
+		offset, size  *uint256.Int
+		expectedError error
+		reduceGas     uint64
+	}{
+		"returns error if memory expansion fails": {
+			offset: uint256.NewInt(math.MaxUint64), size: uint256.NewInt(1),
+			expectedError: errOverflow,
+		},
+		"returns error if word gas cost is more than available gas": {
+			offset: uint256.NewInt(10), size: uint256.NewInt(128),
+			reduceGas:     1,
+			expectedError: errOutOfGas,
+		},
+		"calls emitLog with recipient": {
+			offset: uint256.NewInt(10), size: uint256.NewInt(128),
+		},
+	}
+	for name, test := range tests {
+		for n := 0; n < 4; n++ {
+			t.Run(fmt.Sprintf("%v/LOG%d", name, n), func(t *testing.T) {
+
+				runContext := tosca.NewMockRunContext(gomock.NewController(t))
+				ctxt := getEmptyContext()
+				for i := 0; i < n; i++ {
+					ctxt.stack.push(uint256.NewInt(0))
+				}
+				ctxt.stack.push(test.size)
+				ctxt.stack.push(test.offset)
+				ctxt.memory.store = make([]byte, 10)
+				ctxt.context = runContext
+				ctxt.gas = tosca.Gas(test.size.Uint64()*8 - test.reduceGas)
+				ctxt.params.Recipient = tosca.Address{1}
+				// ignore the expansion error to focus on log operation
+				// this expansion is done to remove expansion costs (if any) and
+				// test word count cost only.
+				_ = ctxt.memory.expandMemory(test.offset.Uint64(), test.size.Uint64(), &context{gas: 1 << 32})
+
+				if test.expectedError == nil {
+					runContext.EXPECT().EmitLog(gomock.Any()).Do(func(log tosca.Log) {
+						if want, got := ctxt.params.Recipient, log.Address; want != got {
+							t.Errorf("unexpected log address, wanted %v, got %v", want, got)
+						}
+						if want, got := n, len(log.Topics); want != got {
```

Comment:
If it is not too much effort: it would be great to check that the right topics are forwarded in the correct order.

I guess the simplest would be to always push the topics 5, 4, 3, 2, 1 on the stack and here make sure that the topics 1-n are present;

Reply:
Added check for topic order and data copy.